lib/Model/events.js

```diff
@@ -221,35 +229,88 @@ Model.prototype.removeContextListeners = function(value) {
   return this;
 };
 
-function eventListener(model, subpattern, cb) {
+/**
+ * @param {Model} model
+ * @param {string} eventType
+ * @param {string | Model} subpattern
+ * @param {Function} cb
+ */
+function eventListener(model, eventType, subpattern, cb) {
   if (cb) {
     // For signatures:
     // model.on('change', 'example.subpath', callback)
     // model.at('example').on('change', 'subpath', callback)
-    var pattern = model.path(subpattern);
-    return modelEventListener(pattern, cb, model._eventContext);
+    var pattern;
+    if (Array.isArray(subpattern)) {
+      pattern = model._splitPath().concat(subpattern);
+    } else {
+      pattern = model.path(subpattern);
+    }
+    return modelEventListener(eventType, pattern, cb, model._eventContext);
   }
   var path = model.path();
-  cb = arguments[1];
+  cb = arguments[2];
   // For signature:
   // model.at('example').on('change', callback)
-  if (path) return modelEventListener(path, cb, model._eventContext);
+  if (path) return modelEventListener(eventType, path, cb, model._eventContext);
   // For signature:
   // model.on('normalEvent', callback)
   return cb;
 }
 
-function modelEventListener(pattern, cb, eventContext) {
-  var patternSegments = util.castSegments(pattern.split('.'));
-  var testFn = testPatternFn(pattern, patternSegments);
-
-  function modelListener(segments, eventArgs) {
-    var captures = testFn(segments);
-    if (!captures) return;
+/**
+ * Returns a function that can be passed to `EventEmitter#on`, with some
+ * additional properties used for `Model#removeAllListeners`.
+ *
+ * When the function is called, it checks if the event matches `patternArg`, and
+ * if there's a match, it calls `cb`.
+ *
+ * @param {string} eventType
+ * @param {string | string[]} patternArg
+ * @param {Function} cb
+ * @param {*} eventContext
+ * @return {ModelListenerFn & ModelListenerProps}
+ */
+function modelEventListener(eventType, patternArg, cb, eventContext) {
```

Comment:
As discussed, we probably also need an explicit option to turn on the structured events in case you are passing in a scoped model instead of a path or array.

Reply:
Good catch!